iocore/nexthop/NextHopHost.h

```diff
@@ -0,0 +1,201 @@
+ *
+ */
+
+// Define a std::hash<>() for the lookmaps
+std_hasher_macro(IpEndpoint, ip, ats_ip_port_hash(ip));
+
+namespace NextHop
+{
+using Mutex = RWLock;
+
+/// uid for a host that should persist between updates and reloads.
+/// It is prefered for other systems to cache HostID instead of a HostNamePtr.
+using HostID = size_t;
+
+struct HostName;
+using HostNamePtr = std::shared_ptr<HostName>;
+
+/// lookup HostID, direct string hash
+HostID getHostID(ts::string_view hostname);
+
+/// lookup HostID with map
+HostID getHostID(IpEndpoint addr);
+
+/// lookup HostNamePtr
+HostNamePtr getHost(HostID);
+HostNamePtr getHost(IpEndpoint addr);
+HostNamePtr getHost(ts::string_view hostname);
+
+/// our internal lookup maps
+// TODO move to .cc
+LookupMap<HostID, HostNamePtr> HostLookupByNameHash;
+LookupMap<IpEndpoint, HostNamePtr> HostLookupByAddr;
+
+/// Defines a host address, status, and a block of associated properties.
+/** Only variables need by TSCore are in this structure.
+ * Plugins should define property blocks.
+ * @see PropBlockDeclare
+ */
+struct HostAddr : public PropertyBlock<HostAddr> {
+  const IpEndpoint m_addr;      ///< ip and port
+  std::atomic_uint m_eol;       ///< "end of life" time, set by DNS time to live
+  std::atomic_bool m_available; ///< true when this IP is available for use
+
+  HostAddr(const IpEndpoint &ip_addr, HostNamePtr host) : m_addr(ip_addr) { HostLookupByAddr.put(ip_addr, host); }
+
+  // Note, this uses PropertyBlock::new and delete to manage.
+
+private:
+  HostAddr();
+};
+
+//////////////////////////////////////////////
+
+/// Defines a host (name), group of IPs, and block of associated properties.
+/** Only variables need by TSCore are in the structure.
+ * Plugins can use property blocks to store host associated data.
+ * @see PropBlockDeclare
+ */
+class HostName : public PropertyBlock<HostName>
+{
+public:
+  // to enforce all references be shared_ptr
+  static HostNamePtr
+  alloc(ts::string_view hostname)
+  {
+    HostID id = std::hash<ts::string_view>()(hostname);
+
+    // get a lock incase we are allocating hosts in parallel
+    LockGuard part_lock(s_lock_pool.getMutex(s_lock_pool.getIndex(id)));
+
+    // return existing host by name, to prevent duplicates
+    HostNamePtr existing_host = getHost(id);
+    if (existing_host) {
+      // increment the count of systems that require this to exist.
+      existing_host->m_strong_ref_count++;
+      return existing_host;
+    }
+
+    HostNamePtr host(new HostName(hostname)); ///< call PropertyBlock::operator new
+    host->propBlockInit();
+
+    HostLookupByNameHash.put(id, host);
+    return host;
+  }
+
+  void
+  free()
+  {
+    ink_assert(m_strong_ref_count > 0); // we are calling 'free' more than 'alloc'
+    ink_assert(hasWriteAccess(this));   // get the lock first, getMutex().lock()
+
+    // decrement the count of systems that require this to exist.
+    m_strong_ref_count--;
+    if (m_strong_ref_count > 0) {
```

Comment:
Why not `if (--m_strong_ref_count > 0) {`?

Reply:
I'm just simple-minded that way.